### PR TITLE
Add Prebid Plugin Renderer

### PR DESCRIPTION
**Pull request by github-richard-depierre**

### CODE REVIEW

**Comment by github-richard-depierre** [comment]
@jsligh This feature is complete and I have implemented it on the internal sample app as an example.

**Comment by YuriyVelichkoPI** [comment]
Hi everyone! Sorry, I really want to help, but very busy this week. If you do not hurry with the release - I'll take a look mid-next week or earlier.

**Comment by bretg** [comment]
This refers to Issue [#697](https://github.com/prebid/prebid-mobile-ios/issues/697)

File: PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRegister.swift [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    private override init() {
+        super.init()
+    }
+    
+    /// Register plugin as renderer
+    @objc public func registerPlugin(_ renderer: PrebidMobilePluginRenderer) {
+        let rendererName = renderer.name
+        
+        queue.async(flags: .barrier) { [weak self] in
+            guard self?.plugins[rendererName] == nil else {
```
Comment (YuriyVelichkoPI) [inline PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRegister.swift:36]: It's hard to follow the condition of this expression. Can we use `if self?.plugins[rendererName] == nil {` ?

File: PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRegister.swift [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+ */
+
+import Foundation
+
+/// Global singleton responsible to store plugin renderer instances
+@objc public class PrebidMobilePluginRegister: NSObject {
+    @objc public static let shared = PrebidMobilePluginRegister()
+    
+    private let queue = DispatchQueue(label: "PrebidMobilePluginRegisterQueue", attributes: .concurrent)
+    private var plugins = [String: PrebidMobilePluginRenderer]()
```
Comment (YuriyVelichkoPI) [inline PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRegister.swift:23]: We need to add unit tests for all functions in this class, especially those that perform sync and async operations on this map.

File: PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRenderer.swift [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+import Foundation
+
+@objc public protocol PrebidMobilePluginRenderer: AnyObject {
+    
+    @objc var name: String { get }
+    @objc var version: String { get }
+    @objc var data: [AnyHashable: Any]? { get }
+    
+    var transactionFactory: PBMTransactionFactory? { get set }
```
Comment (YuriyVelichkoPI) [inline PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRenderer.swift:24]: `PBMTransactionFactory` looks like Prebid SDK's internal class. It doesn't looks like third party renderers shoud utilize it somehow. 

Please explain the purpose of this property in the protocol and how it is used in third-party plugins. Or just remove it if it doesn't make sense for plugins.

File: PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRenderer.swift [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+    
+    /// Creates and returns Banner View for a given Bid Response
+    /// Returns nil in the case of an internal error
+    @objc func createBannerAdView(with frame: CGRect, bid: Bid, adConfiguration: AdUnitConfig, connection: PrebidServerConnectionProtocol, adViewDelegate: (any PBMAdViewDelegate)?)
+    
+    /// Creates and returns an implementation of PrebidMobileInterstitialControllerInterface for a given bid response
+    /// Returns nil in the case of an internal error
+    @objc optional func createInterstitialController( bid: Bid, adConfiguration: AdUnitConfig, connection: PrebidServerConnectionProtocol, adViewManagerDelegate: InterstitialController?, videoControlsConfig: VideoControlsConfiguration?)
```
Comment (YuriyVelichkoPI) [inline PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRenderer.swift:32]: Please format the functions' declaration so the each property is placed on a separate line.

File: PrebidMobile/PrebidMobileRendering/Prebid/PBMCacheRenderers/InterstitialController.swift [hunk 5]
```diff
@@ -64,40 +65,12 @@ public class InterstitialController: NSObject, PBMAdViewManagerDelegate {
     }
     
     @objc public func loadAd() {
```
Comment (YuriyVelichkoPI) [inline PrebidMobile/PrebidMobileRendering/Prebid/PBMCacheRenderers/InterstitialController.swift:67]: Please put a comment for this function

// TODO: provide a more relevant name for this function. 

It's out of the scope of this PR, but I want to return back later and solve the inconvenience of naming the function. Because, in fact, this function doesn't load anything, it initializes the internal class respectively to the received bid and other properties.

File: PrebidMobile/PrebidMobileRendering/Prebid/PBMCacheRenderers/PrebidRenderer.swift [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                adViewDelegate?.failed(toLoad: error)
+                return
+            }
+            self?.displayTransaction(transaction, adConfiguration: adConfiguration, connection: connection, adViewDelegate: adViewDelegate)
+        }
+        
+        PBMWinNotifier.notifyThroughConnection(PrebidServerConnection.shared,
+                                               winning: bid) { [weak self] adMarkup in
+            
+            self?.transactionFactory?.load(withAdMarkup: adMarkup!)
```
Comment (YuriyVelichkoPI) [inline PrebidMobile/PrebidMobileRendering/Prebid/PBMCacheRenderers/PrebidRenderer.swift:55]: Don't use force unwrap

File: PrebidMobile/PrebidMobileRendering/Prebid/PBMCacheRenderers/PrebidRenderer.swift [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            }
+        })
+        
+        PBMWinNotifier.notifyThroughConnection(PrebidServerConnection.shared,
+                                               winning: bid,
+                                               callback: { [weak self] adMarkup in
+            if let adMarkup = adMarkup {
+                self?.transactionFactory?.load(withAdMarkup: adMarkup)
+            } else {
+                //TODO: inform failure
```
Comment (YuriyVelichkoPI) [inline PrebidMobile/PrebidMobileRendering/Prebid/PBMCacheRenderers/PrebidRenderer.swift:107]: Log a respective message instead of a comment

File: Example/PrebidDemo/PrebidDemoObjectiveC/Examples/InApp/CustomRendererDisplayBannerViewController.m [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import "CustomRendererDisplayBannerViewController.h"
+#import "PrebidDemoMacros.h"
+
+NSString * const storedImpDisplayBannerCustomRenderer = @"prebid-demo-banner-320-50";
```
Comment (YuriyVelichkoPI) [inline Example/PrebidDemo/PrebidDemoObjectiveC/Examples/InApp/CustomRendererDisplayBannerViewController.m:19]: What is the purpose of these files? 

From the GitHub UI I don't see changes in the project. 

And as I understand, the integration case for third-party rendering should use `prebid-ita-banner-320-50-meta-custom-renderer` for proper work. 

It also should be an integration case for the Swift demo app.

**Review by YuriyVelichkoPI — CHANGES_REQUESTED** [review]
Sorry for the long silence. 

In general functionality looks good,  but we still need to improve / clean up some places. 

Also it's very important to add unit tests for new classes. 

A new review will be needed once comments are addressed. I promise to do it as soon as fixes are provided.

File: PrebidMobile/PrebidMobileRendering/ORTB/PBMORTBBidRequestExtPrebid.m [hunk 9]
```diff
@@ -40,6 +47,13 @@ - (nonnull PBMJsonDictionary *)toJsonDictionary {
     
     PBMMutableJsonDictionary * const storedRequest = [PBMMutableJsonDictionary new];
     ret[@"storedrequest"] = storedRequest;
+    NSArray *renderers = [[PrebidMobilePluginRegister shared] getAllPlugins];
+//    TODO How to get the isOriginalAPI value here
+//    if (renderers.count != 0 || self.adConfiguration.isOriginalAPI) {
```
Comment (YuriyVelichkoPI) [inline PrebidMobile/PrebidMobileRendering/ORTB/PBMORTBBidRequestExtPrebid.m:52]: `PBMParameterBuilderService` contains `adConfiguration`. 

The most straightforward way is to pass `adConfiguration` to  `[PBMORTBParameterBuilder buildOpenRTBFor:bidRequest];` and further through the call chain.

Comment (YuriyVelichkoPI) [reply PrebidMobile/PrebidMobileRendering/ORTB/PBMORTBBidRequestExtPrebid.m:52]: @github-richard-depierre, just want to remind you that `isOriginalAPI` is not utilized in the logic yet. Please, see my above comment about how to get this value.

File: PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRenderer.swift (hunk 3 above)
Comment (YuriyVelichkoPI) [reply PrebidMobile/PrebidMobileRendering/PluginRenderer/PrebidMobilePluginRenderer.swift:24]: @github-richard-depierre, we should still remove `PBMTransactionFactory` from the public protocol.

**Review by YuriyVelichkoPI — CHANGES_REQUESTED** [review]
Some of the critical comments are not addressed yet.

File: PrebidMobile/PrebidMobileRendering/Prebid/PBMCore/PBMPrebidParameterBuilder.m [hunk 10]
```diff
@@ -301,6 +306,20 @@ - (void)buildBidRequest:(nonnull PBMORTBBidRequest *)bidRequest {
     }
 }
 
++ (nonnull PBMJsonDictionary *)getRenderersJson {
+    PBMMutableJsonDictionary * const sdk = [PBMMutableJsonDictionary new];
+    NSMutableArray *renderersArray = [NSMutableArray array];
+    NSArray *renderers = [[PrebidMobilePluginRegister shared] getAllPlugins];
+    for (id<PrebidMobilePluginRenderer> renderer in renderers) {
```
Comment (YuriyVelichkoPI) [inline PrebidMobile/PrebidMobileRendering/Prebid/PBMCore/PBMPrebidParameterBuilder.m:313]: SDK should send only custom renderers in the request. 

I see two renderers in the request : 

```
"renderers": [{
    "name": "SampleCustomRenderer",
    "version": "1.0.0"
}, {
    "name": "PrebidRenderer",
    "version": "2.2.3"
}]
```

**Review by YuriyVelichkoPI — CHANGES_REQUESTED** [review]

**Review by YuriyVelichkoPI — APPROVED** [review]
LGTM
